tests/__init__.py

```diff
@@ -332,57 +332,101 @@ class PortalMock:
     Parent class for portal tests.
     """
 
-    def __init__(self, session_bus, portal_name: str, app_id: str = "org.example.App"):
-        self.bus = session_bus
+    def __init__(self, dbus_test_case, portal_name: str, app_id: str = "org.example.App"):
+        self.test_case = dbus_test_case
         self.portal_name = portal_name
-        self.p_mock = None
         self.xdp = None
         self.portal_interfaces: Dict[str, dbus.Interface] = {}
         self.dbus_monitor = None
         self.app_id = app_id
+        self.busses = {dbusmock.BusType.SYSTEM: {}, dbusmock.BusType.SESSION: {}}
 
     @property
     def interface_name(self) -> str:
         return f"org.freedesktop.portal.{self.portal_name}"
 
     @property
     def dbus_con(self):
-        return self.bus.dbus_con
+        return self.test_case.dbus_con
 
-    def start_impl_portal(self, params=None, portal=None):
-        """
-        Start the impl.portal for the given portal name. If missing,
-        the portal name is derived from the class name of the test, e.g.
-        ``TestFoo`` will start ``org.freedesktop.impl.portal.Foo``.
-        """
-        portal = portal or self.portal_name
-        self.p_mock, self.obj_portal = self.bus.spawn_server_template(
-            template=f"tests/templates/{portal.lower()}.py",
-            parameters=params,
+    @property
+    def dbus_con_sys(self):
+        return self.test_case.dbus_con_sys
+
+    def get_bus_for_template(self, template: str):
+        # we need the bus address from the template module
+        module = dbusmock.mockobject.load_module(template)
+
+        bustype = dbusmock.BusType.SYSTEM if module.SYSTEM_BUS else dbusmock.BusType.SESSION
+        assert bustype is not None
+
+        if module.BUS_NAME in self.busses[bustype]:
+            return self.busses[bustype][module.BUS_NAME]
+
+        server = dbusmock.SpawnedMock.spawn_for_name(
+            module.BUS_NAME,
+            "/dbusmock",
+            dbusmock.OBJECT_MANAGER_IFACE,
+            bustype,
             stdout=subprocess.PIPE,
         )
-        flags = fcntl.fcntl(self.p_mock.stdout, fcntl.F_GETFL)
-        fcntl.fcntl(self.p_mock.stdout, fcntl.F_SETFL, flags | os.O_NONBLOCK)
-        self.mock_interface = dbus.Interface(self.obj_portal, dbusmock.MOCK_IFACE)
 
-        self.start_dbus_monitor()
+        flags = fcntl.fcntl(server.process.stdout, fcntl.F_GETFL)
+        fcntl.fcntl(server.process.stdout, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+
+        self.busses[bustype][module.BUS_NAME] = server
```

Comment:
Hooray, github doesn't let me do suggestions here... Anwyay, more pythonic may be:
```python
try:
    return self.busses[bustype][module.BUS_NAME]
except KeyError:
    server = ....
    return server
```

Reply:
Reworked this commit a bit. Would be good if you can take another look.